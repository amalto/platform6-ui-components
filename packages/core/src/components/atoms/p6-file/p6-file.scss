@import 'src/styles/atoms/control';

:host {
  box-sizing: border-box;
  display: block;
  position: relative;

  $file-border-color: $border !default;
  $file-radius: $radius !default;

  $file-cta-background-color: $scheme-main-ter !default;
  $file-cta-color: $text !default;
  $file-cta-hover-color: $text-strong !default;
  $file-cta-active-color: $text-strong !default;

  $file-name-border-color: $border !default;
  $file-name-border-style: solid !default;
  $file-name-border-width: 1px 1px 1px 0 !default;
  $file-name-max-width: 16em !default;

  $file-disabled-background-color: $scheme-main !default;
  $file-disabled-border-color: $border !default;
  $file-disabled-shadow: none !default;
  $file-disabled-opacity: 0.5 !default;

  $file-colors: $colors !default;

  .file-cta,
  .file-name {
    @include control;

    border-color: $file-border-color;
    border-radius: $file-radius;
    font-size: 1em;
    padding-left: 1em;
    padding-right: 1em;
    white-space: nowrap;
  }
  .file-cta {
    background-color: $file-cta-background-color;
    color: $file-cta-color;
  }

  .file-name {
    border-color: $file-name-border-color;
    border-style: $file-name-border-style;
    border-width: $file-name-border-width;
    display: block;
    max-width: $file-name-max-width;
    overflow: hidden;
    text-align: inherit;
    text-overflow: ellipsis;
  }
  .file-icon {
    align-items: center;
    display: flex;
    height: 1em;
    justify-content: center;
    @include ltr-property('margin', 0.5em);

    width: 1em;
    .fa {
      font-size: 14px;
    }
  }

  .file-label,
  .file-cta {
    width: 100%;
  }

  .file-input {
    height: 100%;
    left: 0;
    opacity: 0;
    outline: none;
    position: absolute;
    top: 0;
    width: 100%;
  }

  .file {
    @include unselectable;

    align-items: stretch;
    display: flex;
    justify-content: flex-start;
    position: relative;

    // Colors;
    @each $name, $pair in $file-colors {
      $color: nth($pair, 1);
      $color-invert: nth($pair, 2);
      &.is-#{$name} {
        .file-cta {
          background-color: $color;
          border-color: transparent;
          color: $color-invert;
        }
        &:hover {
          .file-cta {
            background-color: bulmaDarken($color, 2.5%);
            border-color: transparent;
            color: $color-invert;
          }
        }
        &:focus {
          .file-cta {
            border-color: transparent;
            color: $color-invert;

            &:not(:active) {
              box-shadow: 0 0 0.5em bulmaRgba($color, 0.25);
            }
          }
        }
        &:active {
          .file-cta {
            background-color: bulmaDarken($color, 5%);
            border-color: transparent;
            color: $color-invert;
          }
        }

        &.is-disabled,
        fieldset[disabled] & {
          .file-cta {
            background-color: $color;
            border-color: transparent;
            box-shadow: none;
          }
        }

        &.is-outlined {
          .file-cta {
            background-color: transparent;
            border-color: $color;
            color: $color;
          }

          &:hover,
          &:focus {
            .file-cta {
              background-color: $color;
              border-color: $color;
              color: $color-invert;
            }
          }

          &.is-disabled,
          fieldset[disabled] & {
            .file-cta {
              background-color: transparent;
              border-color: $color;
              box-shadow: none;
              color: $color;
            }
          }
        }
      }
    }

    // Sizes;
    &.is-small {
      font-size: $size-small;
    }
    &.is-medium {
      font-size: $size-medium;
      .file-icon {
        .fa {
          font-size: 21px;
        }
      }
    }

    &.is-large {
      font-size: $size-large;
      .file-icon {
        .fa {
          font-size: 28px;
        }
      }
    }

    .file-label {
      align-items: stretch;
      display: flex;
      cursor: pointer;
      justify-content: flex-start;
      overflow: hidden;
      position: relative;
    }

    &.is-disabled {
      .file-cta,
      .file-label {
        cursor: not-allowed;
      }

      .file-cta {
        background-color: $file-disabled-background-color;
        border-color: $file-disabled-border-color;
        box-shadow: $file-disabled-shadow;
        opacity: $file-disabled-opacity;
      }
    }

    .file-label:hover(:not(.is-disabled)) {
      .file-cta {
        background-color: bulmaDarken($file-cta-background-color, 2.5%);
        color: $file-cta-hover-color;
      }
      .file-name {
        border-color: bulmaDarken($file-name-border-color, 2.5%);
      }
    }
    .file-label:active(:not(.is-disabled)) {
      .file-cta {
        background-color: bulmaDarken($file-cta-background-color, 5%);
        color: $file-cta-active-color;
      }
      .file-name {
        border-color: bulmaDarken($file-name-border-color, 5%);
      }
    }
  }
}
