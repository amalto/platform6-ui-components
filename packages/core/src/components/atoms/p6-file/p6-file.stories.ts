import { Components } from '../../../components';
import { isCustomEvent } from '../../../shared/form/event';
import { ComponentProps, getElement, getForm, makeModeStory, makeStory, Props } from '../../../shared/storybook';
import { Mode } from '../../../shared/types';

const component = 'p6-file';

export default {
  title: 'Atoms/File',
  component,
};

const componentProps: ComponentProps = ['name', 'placeholder', 'required', 'value'];

const getStoryField = (props?: Props<Components.P6File>): HTMLElement => {
  const elmt = getElement(component, [], props);
  elmt.addEventListener('p6Change', (e: Event): void => {
    if (!isCustomEvent(e)) {
      return;
    }
    // eslint-disable-next-line no-console
    console.log(
      `%c Platform 6 %c ${String.fromCodePoint(0x1f4e1)} %c p6Change :: detail `,
      'background:#61a653; color:white',
      'background:transparent; color:black',
      'background:black; color:cyan',
      e.detail,
    );
  });

  return elmt;
};

export const Default = makeStory<{
  required: boolean;
  placeholder: string;
  value: string;
}>({
  componentProps,
  args: {
    required: false,
    placeholder: 'Placeholder',
    value: '',
  },
  argTypes: {},
  builder: (args): HTMLElement => getStoryField(args),
});

export const Disabled = makeStory<{ disabled: boolean }>({
  componentProps,
  args: { disabled: true },
  builder: (props): HTMLElement => getStoryField({ placeholder: 'Disabled', ...props }),
});

export const Modes = makeModeStory({
  componentProps,
  builder: ({ key, value }) =>
    getStoryField({
      mode: value,
      placeholder: `Mode : ${key}`,
    }),
});

export const Outlined = makeStory<{ outlined: boolean; mode: Mode }>({
  componentProps,
  args: {
    outlined: true,
    mode: Mode.primary,
  },
  builder: (props): HTMLElement => getStoryField({ placeholder: 'Outlined', ...props }),
});

export const Form = makeStory<{
  disabled: boolean;
  required: boolean;
  value: string;
}>({
  componentProps,
  args: {
    disabled: false,
    required: false,
    value: 'default value',
  },
  argTypes: {},
  builder: (props): HTMLElement => getForm(getStoryField({ placeholder: 'Choose a file ...', name: 'field', ...props })),
});
