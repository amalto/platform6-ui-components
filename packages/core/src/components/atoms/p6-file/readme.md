# p6-input

<!-- Auto Generated Below -->

## Properties

| Property                   | Attribute     | Description                                                                                                                                                                | Type                                                                                       | Default        |
| -------------------------- | ------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------ | -------------- |
| `accept`                   | `accept`      | A string that defines the file types the file input should accept. This string is a comma-separated list of filename extension or MIME type example : ".jpg,.png,image/\*" | `string \| undefined`                                                                      | `undefined`    |
| `disabled`                 | `disabled`    | the input is not available for interaction. The value will not be submitted with the form                                                                                  | `boolean`                                                                                  | `false`        |
| `mode`                     | `mode`        | set the mode of the button                                                                                                                                                 | `Mode.danger \| Mode.default \| Mode.info \| Mode.primary \| Mode.success \| Mode.warning` | `Mode.default` |
| `name` _(required)_        | `name`        | The name of the input.                                                                                                                                                     | `string`                                                                                   | `undefined`    |
| `outlined`                 | `outlined`    | Outlined                                                                                                                                                                   | `boolean`                                                                                  | `false`        |
| `placeholder` _(required)_ | `placeholder` | content to be appear in the form control when the form control is empty                                                                                                    | `string`                                                                                   | `undefined`    |
| `required`                 | `required`    | marks an element that can't be submitted without a value.                                                                                                                  | `boolean`                                                                                  | `false`        |
| `size`                     | `size`        | set the size of the button                                                                                                                                                 | `Size.normal \| Size.small`                                                                | `Size.normal`  |
| `value`                    | `value`       | the value of the input.                                                                                                                                                    | `string \| undefined`                                                                      | `undefined`    |

## Events

| Event              | Description                                                                            | Type                                         |
| ------------------ | -------------------------------------------------------------------------------------- | -------------------------------------------- |
| `p6Change`         | Fires after the selection of the files                                                 | `CustomEvent<File[]>`                        |
| `p6FormRegister`   | Registering the field in a p6-form                                                     | `CustomEvent<P6Control<P6FileValue>>`        |
| `p6FormUnregister` | Unregistering the field in a p6-form                                                   | `CustomEvent<P6Control<P6FileValue>>`        |
| `p6Invalid`        | Fires when the field has been checked for validity and doesn't satisfy its constraints | `CustomEvent<InvalidEventDetail>`            |
| `p6Valid`          | Fires when the field has been checked for validity and satisfy its constraints         | `CustomEvent<ValidEventDetail<P6FileValue>>` |

## Methods

### `checkValidity() => Promise<boolean>`

Returns whether a form will validate when it is submitted, without having to submit it.

#### Returns

Type: `Promise<boolean>`

### `reset() => Promise<boolean>`

Restores the input's default value

#### Returns

Type: `Promise<boolean>`

### `validationMessage() => Promise<string>`

Returns the error message that would be displayed if the user submits the form, or an empty string if no error message.
It also triggers the standard error message, such as "this is a required field".

#### Returns

Type: `Promise<string>`

---

_Built with [StencilJS](https://stenciljs.com/)_
