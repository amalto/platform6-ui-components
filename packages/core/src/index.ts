export { Components, JSX } from './components';
export { P6ButtonType } from './components/atoms/p6-button/p6-button';
export { P6CalendarType } from './components/atoms/p6-calendar/p6-calendar';
export { P6CheckboxValue } from './components/atoms/p6-checkbox/p6-checkbox';
export { P6InputValue } from './components/atoms/p6-input/p6-input';
export { P6InputType } from './components/atoms/p6-input/types';
export { Target } from './components/atoms/p6-link/p6-link';
export { P6RadioValue } from './components/atoms/p6-radio/p6-radio';
export { P6SelectControl, P6SelectValue } from './components/atoms/p6-select-native/p6-select-native';
export { P6SwitchValue } from './components/atoms/p6-switch/p6-switch';
export { P6TextareaValue } from './components/atoms/p6-textarea/p6-textarea';
export { P6LanguageValue } from './components/molecules/p6-language/p6-language';
export { P6TranslationControl, P6TranslationValue } from './components/molecules/p6-translation/p6-translation';
export {
  AlignColumnDetail,
  CellValueChangedDetail,
  Column,
  ColumnDefinition,
  ColumnId,
  DataItem,
  EditingCellDetail,
  FilterRowsDetail,
  MoveColumnDetail,
  ResetDefinitionsDetail,
  ResizeColumnDetail,
  Row,
  RowId,
  ShowColumnDetail,
  ShowOptionsDetail,
  SortColumnDetail,
} from './components/organisms/p6-grid/core/entities';
export { ServiceStatus } from './components/templates/p6-service-panel/service-status.enum';
export { P6Control } from './shared/form/control';
export { InvalidEventDetail, isCustomEvent, isInvalidEvent, isValidEvent, ValidEventDetail } from './shared/form/event';
export { Alignment, Mode, modes, Position, positions, Size, sizes, SortOrder } from './shared/types';
export { EnumArrayEntry, enumArrayToObject, enumToArray, EnumType } from './utils/enum';
export { LanguageCode } from './utils/language';
