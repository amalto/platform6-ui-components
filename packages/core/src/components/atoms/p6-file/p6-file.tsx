import { icon } from '@fortawesome/fontawesome-svg-core';
import { Component, ComponentInterface, Element, Event, EventEmitter, h, Host, JSX, Method, Prop, State } from '@stencil/core';
import { P6Control } from '../../../shared/form/control';
import { InvalidEventDetail, ValidEventDetail } from '../../../shared/form/event';
import { defaultCheckValidity, defaultValidationMessage } from '../../../shared/form/validation';
import { Mode, Size } from '../../../shared/types';
import { noop } from '../../../utils';
import { cleanupValue, isEmpty } from '../../../utils/attribute';
import { getModeClass, getSizeClass } from '../../../utils/classes';

export type P6FileValue = File[] | undefined;

@Component({
  tag: 'p6-file',
  styleUrl: 'p6-file.scss',
  shadow: true,
})
export class P6File implements ComponentInterface, P6Control<P6FileValue> {
  @Element() host!: HTMLP6FileElement;

  /**
   * A string that defines the file types the file input should accept.
   * This string is a comma-separated list of filename extension or MIME type
   * example : ".jpg,.png,image/*"
   */
  @Prop() accept: string | undefined;

  /**
   * the input is not available for interaction. The value will not be submitted with the form
   */
  @Prop() disabled = false;

  /**
   * set the mode of the button
   */
  @Prop() mode: Mode = Mode.default;

  /**
   * The name of the input.
   */
  @Prop() name!: string;

  /**
   * Outlined
   */
  @Prop() outlined = false;

  /**
   * content to be appear in the form control when the form control is empty
   */
  @Prop() placeholder!: string;

  /**
   * marks an element that can't be submitted without a value.
   */
  @Prop() required = false;

  /**
   * set the size of the button
   */
  @Prop() size: Size = Size.normal;

  /**
   * the value of the input.
   */
  @Prop() value: string | undefined;

  @State() hasError = false;

  /**
   * Registering the field in a p6-form
   */
  @Event() p6FormRegister!: EventEmitter<P6Control<P6FileValue>>;

  /**
   * Unregistering the field in a p6-form
   */
  @Event() p6FormUnregister!: EventEmitter<P6Control<P6FileValue>>;

  /**
   * Fires when the field has been checked for validity and satisfy its constraints
   */
  @Event() p6Valid!: EventEmitter<ValidEventDetail<File[] | undefined>>;

  /**
   * Fires when the field has been checked for validity and doesn't satisfy its constraints
   */
  @Event() p6Invalid!: EventEmitter<InvalidEventDetail>;

  /**
   * Fires after the selection of the files
   */
  @Event() p6Change!: EventEmitter<File[]>;

  private nativeInput: HTMLInputElement | undefined;

  private defaultValue: string | number | undefined;

  componentWillLoad(): void {
    if (this.defaultValue === undefined) {
      this.defaultValue = this.value;
    }
  }

  render(): JSX.Element {
    const containerClass = {
      'file': true,
      ...getModeClass(this.mode),
      ...getSizeClass(this.size),
      'is-outlined': this.outlined,
      'is-disabled': this.disabled,
    };

    const classes = {
      'file-input': true,
      'is-danger': this.hasError,
    };
    const img = icon({ prefix: 'fas', iconName: 'upload' });

    return (
      <Host>
        <form noValidate class={containerClass}>
          <label class="file-label">
            <input
              class={classes}
              ref={(input): void => {
                this.nativeInput = input;
              }}
              accept={this.accept}
              disabled={this.disabled}
              name={this.name}
              required={this.required}
              type="file"
              onChange={this.onChangeHandler()}
            />
            <span class="file-cta">
              <span class="file-icon" innerHTML={img.html[0]} />
              <span class="file-label">{cleanupValue(this.placeholder)}</span>
            </span>
          </label>
        </form>
      </Host>
    );
  }

  async componentDidLoad(): Promise<void> {
    if (!isEmpty(this.value)) {
      await this.checkValidity();
    }

    this.p6FormRegister.emit(this);
  }

  disconnectedCallback?(): void {
    this.p6FormUnregister.emit(this);
  }

  /**
   * Returns the error message that would be displayed if the user submits the form, or an empty string if no error message.
   * It also triggers the standard error message, such as "this is a required field".
   */
  @Method()
  async validationMessage(): Promise<string> {
    return defaultValidationMessage(this.nativeInput);
  }

  /**
   * Returns whether a form will validate when it is submitted, without having to submit it.
   */
  @Method()
  async checkValidity(): Promise<boolean> {
    return defaultCheckValidity<File[]>({
      name: this.name,
      disabled: false,
      nativeInput: this.nativeInput,
      p6Valid: this.p6Valid,
      p6Invalid: this.p6Invalid,
      validationMessage: this.validationMessage.bind(this),
      errorHandler: hasError => {
        this.hasError = hasError;
      },
      getValue: (): File[] => this.files,
    });
  }

  /**
   * Restores the input's default value
   */
  @Method()
  async reset(): Promise<boolean> {
    let result = false;

    if (this.nativeInput) {
      this.nativeInput.value = this.defaultValue !== undefined ? this.defaultValue.toString() : '';
      result = true;
    }

    return Promise.resolve(result);
  }

  private get files(): File[] {
    return this.nativeInput?.files === undefined || this.nativeInput?.files === null ? [] : Array.from(this.nativeInput.files);
  }

  private onChangeHandler(): (event: Event) => void {
    return event => {
      event.stopPropagation();
      event.preventDefault();

      this.checkValidity()
        .then(isChecked => {
          const isValid = isChecked && this.nativeInput?.validationMessage === '';
          if (isValid) {
            this.p6Change.emit(this.files);
          }
          return isValid;
        })
        .catch(noop);
    };
  }
}
