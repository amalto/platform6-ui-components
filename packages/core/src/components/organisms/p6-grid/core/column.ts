import { Alignment, Direction, SortOrder } from '../../../../shared/types';
import { throwError } from '../../../../utils/error';
import { Column, ColumnDefinition, ColumnId, DataItem } from './entities';

export const MIN_WIDTH = 75;
export const DEFAULT_WIDTH = MIN_WIDTH;
export const INC_WIDTH = 10;

export function defaultGetValue<T extends DataItem>(data: T, context: ColumnDefinition<T>): string {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'getValue' function for ${JSON.stringify(context)}`);
  }

  return `${data[context.field as keyof T]}`;
}

export function defaultSetValue<T extends DataItem>(value: string, data: T, context: ColumnDefinition<T>): T {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'setValue' function for ${JSON.stringify(context)}`);
  }

  return {
    ...data,
    [context.field as keyof T]: value,
  };
}

let columnIds = 0;
export function fromDefinition<T extends DataItem>(column: ColumnDefinition<T>): Column<T> {
  // eslint-disable-next-line no-plusplus
  const id = column.id || column.field || columnIds++;

  return {
    ...column,
    id: `col-${id}`,
    editable: column.editable || false,
    hidden: column.hidden || false,
    width: column.width || 100,
    align: column.align || Alignment.center,
    sortOrder: column.sortOrder || SortOrder.none,
    disableHeaderMenu: column.disableHeaderMenu || false,
    getValue: column.getValue || defaultGetValue,
    setValue: column.setValue || defaultSetValue,
  };
}

export function isHidden<T extends DataItem>(column: Column<T>): boolean {
  return !!column.hidden;
}

export function replaceColumn<T extends DataItem>(updatedColumn: Column<T>, allColumns: Column<T>[]): Column<T>[] {
  return allColumns.map(col => (col.id === updatedColumn.id ? updatedColumn : col));
}

export function move<T extends DataItem>(id: ColumnId, direction: Direction, columns: Column<T>[]): Column<T>[] {
  const from = columns.findIndex(col => col.id === id);
  const to = from + (direction === Direction.left ? -1 : 1);

  if (from < 0 || to < 0 || to >= columns.length) {
    return columns;
  }

  const result = columns.slice();
  const item = result.splice(from, 1);

  return result.slice(0, to).concat(item).concat(result.slice(to, columns.length));
}
