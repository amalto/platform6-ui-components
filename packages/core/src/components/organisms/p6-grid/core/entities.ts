import { EventEmitter, JSX } from '@stencil/core';
import { Alignment, Direction, SortOrder } from '../../../../shared/types';

export type ColumnId = string;

export type DataItem = Record<string, unknown>;

export interface ColumnDefinition<T extends DataItem> {
  id?: ColumnId;
  label: string;
  field?: string | keyof T;
  editable?: boolean;
  filtreable?: boolean;
  sortable?: boolean;
  hidden?: boolean;
  width?: number;
  align?: Alignment;
  color?: string;
  sortOrder?: SortOrder;
  disableHeaderMenu?: boolean;
  getValue?: (data: T, context: ColumnDefinition<T>) => string;
  setValue?: (value: string, data: T, context: ColumnDefinition<T>) => T;
  cellEditor?: (data: T, context: ColumnDefinition<T>) => JSX.Element & CellEditor<T>;
  cellRenderer?: (data: T, rowId: RowId, context: ColumnDefinition<T>) => string | HTMLElement;
}

export interface Column<T extends DataItem> extends ColumnDefinition<T> {
  id: ColumnId;
  editable: boolean;
  hidden: boolean;
  width: number;
  align: Alignment;
  sortOrder: SortOrder;
  disableHeaderMenu: boolean;
  getValue: (data: T, context: ColumnDefinition<T>) => string;
  setValue: (value: string, data: T, context: ColumnDefinition<T>) => T;
}

export interface CellEditor<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  value: string;

  p6GridCellValueChanged: EventEmitter<CellValueChangedDetail<T>>;
  p6GridEditingCell: EventEmitter<EditingCellDetail<T>>;
}
export type RowId = string;

export interface Row<T extends DataItem> {
  id: RowId;
  data: T;
}

interface ColumnEventDetail<T extends DataItem> {
  column: Column<T>;
}

export interface MoveColumnDetail<T extends DataItem> extends ColumnEventDetail<T> {
  direction: Direction;
}
export type ResizeColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type AlignColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type SortColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type ShowColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export interface CellValueChangedDetail<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  oldValue: string;
}

export enum EditingCellStatus {
  Start,
  Stop,
  Cancel,
}

export interface EditingCellDetail<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  status: EditingCellStatus;
}

export interface ResetDefinitionsDetail {
  reset: boolean;
}

export interface ShowOptionsDetail {
  visible: boolean;
}

export interface FilterRowsDetail {
  value: string;
}
