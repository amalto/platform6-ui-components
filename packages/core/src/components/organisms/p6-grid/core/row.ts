import { SortOrder } from '../../../../shared/types';
import { Column, DataItem, Row, RowId } from './entities';

let rowIds = 0;
export function fromData<T extends DataItem>(data: T): Row<T> {
  rowIds += 1;

  return {
    id: `row-${rowIds}`,
    data,
  };
}

export function compareRow<T extends DataItem>(column: Column<T> | undefined): (a: Row<T>, b: Row<T>) => number {
  const sortOrder = column?.sortOrder || SortOrder.none;
  const factor = sortOrder === SortOrder.desc ? -1 : 1;

  return (a: Row<T>, b: Row<T>): number => {
    if (column === undefined || column.sortable === false || sortOrder === SortOrder.none) {
      return 0;
    }

    const valueA = column.getValue(a.data, column);
    const valueB = column.getValue(b.data, column);

    return valueA.localeCompare(valueB) * factor;
  };
}

export function filterBy<T extends DataItem>(searchValue: string, columns: Column<T>[]): (row: Row<T>) => boolean {
  return row => columns.filter(col => col.filtreable !== false).some(col => col.getValue(row.data, col).includes(searchValue));
}

export function rangeSelectRow<T extends DataItem>(currentRowIdSelect: RowId, selectedRows: Set<RowId>, displayedRow: Row<T>[]): Set<RowId> {
  const currentSelectedIndex = displayedRow.findIndex(row => row.id === currentRowIdSelect);

  if (currentSelectedIndex === -1) {
    return selectedRows;
  }

  const previousRowIdSelected = Array.from(selectedRows)
    .reverse()
    .find(rowId => displayedRow.findIndex(row => row.id === rowId) > -1);

  if (previousRowIdSelected === undefined) {
    return new Set(selectedRows.add(currentRowIdSelect));
  }

  const previousSelectedIndex = displayedRow.findIndex(row => row.id === previousRowIdSelected);

  let [from, to] = [previousSelectedIndex, currentSelectedIndex];
  let excludedFromSlice = currentRowIdSelect;
  if (currentSelectedIndex < previousSelectedIndex) {
    [from, to] = [to, from];
    excludedFromSlice = previousRowIdSelected;
  }

  return displayedRow
    .slice(from, to)
    .reduce((selected, cur) => {
      return cur.id === undefined ? selected : selected.add(cur.id);
    }, new Set(selectedRows))
    .add(excludedFromSlice);
}

export function replaceRow<T extends DataItem>(updatedRow: Row<T>, allRows: Row<T>[]): Row<T>[] {
  return allRows.map(row => (row.id === updatedRow.id ? updatedRow : row));
}
