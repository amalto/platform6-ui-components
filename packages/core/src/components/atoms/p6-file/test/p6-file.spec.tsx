import { library } from '@fortawesome/fontawesome-svg-core';
import { faUpload } from '@fortawesome/free-solid-svg-icons';
import { newSpecPage } from '@stencil/core/testing';
import { P6File } from '../p6-file';

library.add(faUpload);

describe('p6-file', () => {
  Object.defineProperty(HTMLInputElement.prototype, 'checkValidity', {
    writable: true,
    value: jest.fn().mockImplementation(() => true),
  });

  it('renders', async () => {
    const page = await newSpecPage({
      components: [P6File],
      html: `<p6-file></p6-file>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
