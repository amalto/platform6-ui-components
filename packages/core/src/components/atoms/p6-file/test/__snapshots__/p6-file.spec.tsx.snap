// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-file renders 1`] = `
<p6-file>
  <mock:shadow-root>
    <form class="file" novalidate="">
      <label class="file-label">
        <input class="file-input" type="file">
        <span class="file-cta">
          <span class="file-icon">
            <svg aria-hidden="true" class="fa-upload fa-w-16 svg-inline--fa" data-icon="upload" data-prefix="fas" focusable="false" role="img" viewBox="0 0 512 512" xmlns="http://www.w3.org/2000/svg">
              <path d="M296 384h-80c-13.3 0-24-10.7-24-24V192h-87.7c-17.8 0-26.7-21.5-14.1-34.1L242.3 5.7c7.5-7.5 19.8-7.5 27.3 0l152.2 152.2c12.6 12.6 3.7 34.1-14.1 34.1H320v168c0 13.3-10.7 24-24 24zm216-8v112c0 13.3-10.7 24-24 24H24c-13.3 0-24-10.7-24-24V376c0-13.3 10.7-24 24-24h136v8c0 30.9 25.1 56 56 56h80c30.9 0 56-25.1 56-56v-8h136c13.3 0 24 10.7 24 24zm-124 88c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20zm64 0c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20z" fill="currentColor"></path>
            </svg>
          </span>
          <span class="file-label"></span>
        </span>
      </label>
    </form>
  </mock:shadow-root>
</p6-file>
`;
